### ISSUE TITLE
Models using polymorphic has_many load associations with an integer `<associated>_type` column

### ISSUE THREAD

**Issue by bgracie** [rails/rails#27149]
### Steps to reproduce

https://gist.github.com/bgracie/ad295a4289e79cbcd64ea4fba3c91801

### Expected behavior
When we call a `has_many :as` association, it should not return records that have an integer `<associated>_type` column.  Rather, it should throw an error.

### Actual behavior
If a `has_many :as` association targets a model with an integer `<associated>_type` column, when we call that association it will coerce the class name to 0  and match any targets with 0 in the `<associated>_type` column.

### System configuration
**Rails version**: 5.0

**Ruby version**: 2.2.4

**Comment by matrix9180** [comment]
Not sure that this warrants it's own exception. The documentation states that the _type should be a string column, and I don't believe we have an exception if you go against the docs and for example make an enum from a string column instead of integer.

**Comment by bgracie** [comment]
Sure, I appreciate that you guys have to make a policy decision about whether you want to code exceptions around people not following what's in the docs.  All I can offer as input for this decision is my own experience with this problem:

- I'm generally familiar with how to set up polymorphic associations.
- Either my colleague or I made the `_type` column as an `int` as a careless mistake -- we weren't trying to get creative or anything.
- The feature appeared to work because AR dragged up records that matched the ID, and there weren't many conflicting ID's till later.  We also never had any reason to call the association method on the table with the polymorphic columns (which would have triggered `NoMethodError: undefined method 'constantize' for 0:Fixnum`) 
- As a result of the feature apparently working, we ended up with hundreds of records with ambiguous joins (where both join tables had records with the appropriate ID).  As a result, we spent roughly 10 dev hours programmatically making 'best guesses' to assign proper `_type` field values and creating a system by which those guesses could be reviewed.  And then there will probably be many admin hours spent manually reviewing these cases to make sure those guesses were right.  

TLDR: it was a costly and embarrassing bug, and I filed this report because I hope nobody else gets burned in a similar way.

**Comment by matrix9180** [comment]
Fair enough. I'm not on the core team so I have no final say anyway, just a long time user trying to help out on some issues. 💯 

Started looking through the code to see if I could add a warning when the _type isn't a string but the AR codebase is quite a beast and I'm having trouble wrapping my head around all of it at the moment. 

I do agree there should at least be some kind of warning when this happens, though.

**Comment by matrix9180** [comment]
As for fixing your existing data, I've been there quite a few times, having to try to piece things back together from (not verbose enough) logs. 
In new migrations you can use t.belongs_to :foo, polymorphic: true and it will automatically create the foo_type, or you can do foo:belongs_to{polymorphic}  with the generator. Might help avoid it in the future.

**Comment by bgracie** [comment]
Cool, thanks for the advice.  If the core team decides to proceed with the warning, I can try to find some time to implement it.  Though...like you, I did have a look at AR's guts and was completely lost!

**Comment by sgrif** [comment]
I don't think this is a common enough issue to warrant code specifically for it.

**Comment by sgrif** [comment]
I would consider a PR if it was low complexity and didn't look like it'd have a high maintenance burden, though.
